server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/FrequencyCappedAction.java

```diff
@@ -21,15 +20,16 @@ public class FrequencyCappedAction {
     private final LongSupplier currentTimeMillisSupplier;
     private TimeValue minInterval;
 
-    private long next = -1;
+    private long next;
 
-    public FrequencyCappedAction(ThreadPool threadPool) {
-        this(threadPool.relativeTimeInMillisSupplier());
+    public FrequencyCappedAction(LongSupplier currentTimeMillisSupplier) {
+        this(currentTimeMillisSupplier, TimeValue.ZERO);
     }
 
-    public FrequencyCappedAction(LongSupplier currentTimeMillisSupplier) {
+    public FrequencyCappedAction(LongSupplier currentTimeMillisSupplier, TimeValue initialDelay) {
         this.currentTimeMillisSupplier = currentTimeMillisSupplier;
         this.minInterval = TimeValue.MAX_VALUE;
+        this.next = currentTimeMillisSupplier.getAsLong() + initialDelay.getMillis();
```

Comment:
This approach will work if:
* new node is added
* (AND) it is elected as master and is handling reconciliation.

however I am afraid more common scenario is:
* a new (data) node is added
* existing elected master handles reconciliation. Since it is initialized long time ago, it keeps emitting warnings as is.

Since we modify `FrequencyCappedAction` that already have some state I wonder if we should do something like:
```
   private long delayedNotification = -1

    public void maybeExecute(Runnable runnable) {
        var current = currentTimeMillisSupplier.getAsLong();
        if (delayedNotification == -1) {
            delayedNotification = current + initialDelay;
        } else if (current > delayedNotification) {
            runnable.run();
            delayedNotification = current + minInterval.millis();
        }
    }

    // should be called if/when we detect the sate return back to normal
    public void executionNoLongerNeeded() {
            delayedNotification = -1;
    }
```

Reply:
This solution does target the "more common scenario" as you described. Note that `FrequencyCappedAction#next` starts from `-1`. 
https://github.com/elastic/elasticsearch/blob/4f62da96dd7bca65b6c79ebe855083ddf6ad4b27/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/FrequencyCappedAction.java#L24

This means the first warning message will always be emitted. The noise that we are seeing on the logging cluster is like the follows:
1. A master node and a search node started with about 50 shards
2. A second search node started after 1 min. 
3. The master node decided to move almost half of the shards to the new search node and logged the warning for undesirable allocation.

The above scenario happen within ~3 min, which is the reason I chose 5 min to be the initial delay. I think it should work for the particalur situation described above.

That said, it is indeed a limited solution. It does _not_ work for either:
1. If the 2nd search node joins after 5 min
2. Or if the new node joins after a warning is already emitted (i.e. initialDelay is already used) and after the `minInterval`. 

We don't currently see noises for the above situations. I think this is probably due to number of shards being larger in these situations. That is, a new node can more easily lead to undesired allocation warning if the total number of shards is low, e.g. 50. But it becomes less likely for higher number of shards, e.g. 500. 

I looked into the suggested code. I think it sometimes disables `minInterval`? Assuming `executionNoLongerNeeded` is called each time when `maybeLogUndesiredAllocationsWarning` decides to not log, it sets `delayedNotification` to `-1` which means the next warning log will be emited after `initialDelay` regardless of the `minInterval`?

Reply:
As discussed offline, currently the first scenario is happening a lot, so this is worth merging.
The second scenario might be addressed later separately.